### ISSUE TITLE
Abstract from the typesafe config backend

### ISSUE THREAD

**Issue by melrief** [pureconfig/pureconfig#87]
pureconfig started as layer on top of the typesafe config library. The idea was to keep intact the parsing of files and the resolution rules while adding a generic derivation system that converts a `ConfigValue` to a value of an arbitrary type. The initial idea was to not tie the library to typesafe config and in future releases to allow to plug in other backends. I would like to start a discussion about this and eventually implement this abstraction in our codebase. My thoughts about this:

1. **abstraction**: right now the structure that represents a raw configuration is [`ConfigValue`](https://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigValue.html). `ConfigValue` represents JSON-like values with support for [boolean, list, null, number, object and string](https://typesafehub.github.io/config/latest/api/com/typesafe/config/ConfigValueType.html). If we decide to abstract from the backend, using typesafe structures makes no sense because we don't want to have typesafe config as dependency when it's not required. What I propose it to create our own representation of a JSON-like value and keep it as small and minimalistic as possible. For each backend, a new adapter from the backend specific structure to our representation should be coded and that's the only required piece of code to make the new backend work with pureconfig. I am open to use an external library to represent a configuration if there is a good reason for it.

2. **loading**: there are multiple ways to load a backend implementation at runtime. In Java it is common to set the implementor class somehow, e.g. via configuration files, and then use the class loader to load the specified class. This style allows to create a new backend without recompiling pureconfig but has the side-effect that there are no checks at compile time. I personally don't like this style even if I understand the great flexibility it provides. I like to have my compiler to link all the pieces together before I use the library. Consequently I propose to make the core of pureconfig abstract from the implementation and then have one subproject for each backend. If a user wants to implement a new backend, he/she just have to add a new subproject with minimum effort. Each backend would have its own artifact on maven.

Let me know what you think about these two points and if you want to add other points to the discussion. I would prefer to have a clear idea on which everybody agree before we implement it.

**Comment by leifwickland** [comment]
I believe the current guts of `ConfigConvert` don't really use the `boolean`, `number`, etc parts of the `ConfigValue` interface. `ConfigConvert` type-tests and then extracts a list, object, or string representation of the value. To some degree that couples the type-specific `ConfigConvert` instances to the string representation of data that are produced would be produced by the various backends. I'm not sure what an alternative would look like. 

All of that to say, I'm not sure if the solution to abstraction needs to include any more than:
- get type
- get list
- get map/object
- get string representation

**Comment by mberndt123** [comment]
> What I propose it to create our own representation of a JSON-like value and keep it as small and minimalistic as possible.

Here's something the Scala community doesn't need: another JSON library, there are dozens of them already out there.

And actually, now that I think about it: what value does pureconfig add over circe-config? Because that can also read stuff from typesafe config files.

**Comment by jcazevedo** [comment]
> And actually, now that I think about it: what value does pureconfig add over circe-config? Because that can also read stuff from typesafe config files.

circe-config works by translating a `Config` into circe's `Json`, whereas PureConfig operates directly on `Config`. By converting to circe's `Json`, you lose the file system location of `ConfigValue`s, which are relevant for error reporting. 

From the top of my head, there are also some things that PureConfig does by default which makes it more aligned with the HOCON specification. I think all of these are doable using circe and circe-config, but require some custom code:

* Field names in derived `Reader`s and `Writer`s are translated from `camelCase` to `kebab-case`, the latter being the [de facto standard](https://github.com/lightbend/config/blob/master/HOCON.md#hyphen-separated-vs-camelcase) for HOCON;
* `ConfigCursor` applies the same [automatic type conversions](https://github.com/lightbend/config/blob/master/HOCON.md#automatic-type-conversions) the HOCON specification suggests;
* The `Duration` and `Period` readers respect the HOCON [suggested](https://github.com/lightbend/config/blob/master/HOCON.md#duration-format) [formats](https://github.com/lightbend/config/blob/master/HOCON.md#period-format).

**Comment by mberndt123** [comment]
Right, so all of the stuff you describe seems to be HOCON-specific, which means that abstracting from HOCON would sort of defeat the purpose of the library, right?

**Comment by jcazevedo** [comment]
> Right, so all of the stuff you describe seems to be HOCON-specific, which means that abstracting from HOCON would sort of defeat the purpose of the library, right?

Right. I'm not an advocate for abstracting the config backend in the core module. I prefer to keep the core simple and, if need be, have modules that translate other format's AST to Typesafe Config's, such as the one we have for YAML.

**Comment by mberndt123** [comment]
> Right. I'm not an advocate for abstracting the config backend in the core module. I prefer to keep the core simple and, if need be, have modules that translate other format's AST to Typesafe Config's, such as the one we have for YAML.

To be honest that sounds like a bit of a hack to me. The Typesafe Config library supports HOCON, JSON and .properties files, and the include statement in HOCON allows you to include data in all three formats. Adding YAML support in PureConfig wouldn't allow that, hence I think that additional formats should be added directly to the typesafe config library rather than pureconfig.

And in fact there is an issue for adding YAML support to the typesafe config library:
https://github.com/lightbend/config/issues/494

The problem of course is that they don't want to bloat their library with additional dependencies like a YAML parser, so there would have to be some mechanism to load support for additional formats at run-time. I believe that if you want additional formats, that would be the right way to do it.

**Comment by ekrich** [comment]
You could also consider supporting `sconfig` as a way of potentially supporting Scala.js and Scala Native. https://github.com/ekrich/sconfig/
